File: backend/webhook_sender.py
from datetime import datetime
import json
from typing import Optional, Dict, Any
from pydantic import BaseModel
import os
import hmac
import httpx 
import asyncio
import logging
import generate_secret 
import requests

# Get logger
logger = logging.getLogger('dispatch_logger')

class WebhookSender:
    def __init__(self, base_url: str, secret_key: str):
        self.base_url = base_url
        self.secret_key = secret_key or os.getenv('WEBHOOK_SECRET_KEY')
    
    # Only do this if there's no current secret key. Otherwise ignore. 
    def _generate_secret(self, length=32):
        return generate_secret.generate_secret(length)
    
    def _get_signature(self, payload: Dict[str, Any]):
        return hmac.new(self.secret_key.encode(), payload.encode(), digestmod='sha256').hexdigest()
    
    async def send_webhook(self, 
                           event: str, 
                           data: Dict[str, Any], 
                           retry_count: Optional[int] = 3, 
                           timeout: Optional[float] = 10.0) -> bool:
        payload = {"event": event, "data": data, "timestamp": datetime.now().isoformat()}

        signature = self._get_signature(json.dumps(payload))
        headers = {
            "Content-Type": "application/json",
            "X-Webhook-Signature": signature, 
            "User-Agent": "Fleet Dispatch Assistant"
        }

        for attempt in range(retry_count):
            try: 
                async with httpx.AsyncClient() as client:
                    response = await client.post(self.base_url, json=payload, headers=headers, timeout=timeout)
                    response.raise_for_status()
                    if(response.status_code == 200):
                        logger.info(f"Webhook sent successfully. Status code: {response.status_code}")
                        return True
                    else:
                        logger.error(f"Failed to send webhook. Status code: {response.status_code}")
                        return False
            
            except httpx.HTTPError as e:
                
                if attempt < retry_count - 1:
                    logger.warning(f"Failed to send webhook. Retrying in 5 seconds. Error: {e}")
                    await asyncio.sleep(5)
                else:
                    logger.error(f"Failed to send webhook after {retry_count} attempts. Error: {e}")
                    return False

        return False

    async def query_webhook(self, data: Dict[Any, Any], retry_count: Optional[int] = 3, timeout: Optional[float] = 10.0) -> Dict[Any, Any]:
        url = os.getenv("NEXT_JS_API_URL", "http://localhost:3000/api/")
        
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "Fleet Dispatch Assistant",
            "x-webhook-secret": os.getenv("WEBHOOK_SECRET_KEY")
        }
        
        payload = {
            "event": self.event,
            "data": data
        }

        response = requests.post(url, json=payload, headers=headers)
        if response.status_code == 200:
            return response.json()
        else:
            raise Exception(f"Failed to query webhook. Status code: {response.status_code}; Response: {response.text}")

File: web/fleet-dispatch-web/src/app/api/ai/webhooks/webhookSender.ts
import crypto from 'crypto';
import generateSecret from './generateSecret';

interface WebhookPayload {
    event: string;
    data: any;
    timestamp?: string;
}

interface WebhookResponse {
    success: boolean;
    response?: any;
    error?: string;
}

class WebhookSender {
    private baseUrl: string;
    private secretKey: string;

    constructor(baseUrl?: string, secretKey?: string) {
        this.baseUrl = baseUrl || process.env.FASTAPI_BASE_URL || 'http://localhost:8000/api/ai/receive-webhook';
        this.secretKey = secretKey || process.env.WEBHOOK_SECRET_KEY || '';
    }

    /**
     * Generate a secret if one doesn't exist
     */
    generateSecret(length: number = 32): string {
        return generateSecret(length);
    }

    /**
     * Create HMAC signature for webhook payload
     */
    getSignature(payload: WebhookPayload | string): string {
        const payloadString = typeof payload === 'string' ? payload : JSON.stringify(payload);
        return crypto
            .createHmac('sha256', this.secretKey)
            .update(payloadString)
            .digest('hex');
    }

    /**
     * Send webhook to FastAPI server Might not be needed or YES? lol 
     */
    async sendWebhook(
        event: string, 
        data: any, 
        retryCount: number = 3, 
        timeout: number = 10000
    ): Promise<WebhookResponse> {
        const payload: WebhookPayload = {
            event,
            data,
            timestamp: new Date().toISOString()
        };

        const signature = this.getSignature(payload);
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'X-Webhook-Signature': signature,
            'User-Agent': 'Fleet Dispatch Assistant'
        };

        for (let attempt = 0; attempt < retryCount; attempt++) {
            try {
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), timeout);

                const response = await fetch(this.baseUrl, {
                    method: 'POST',
                    headers,
                    body: JSON.stringify(payload),
                    signal: controller.signal
                });

                clearTimeout(timeoutId);

                if (response.ok) {
                    console.log(`Webhook sent successfully. Status: ${response.status}`);
                    return { success: true, response: await response.json() };
                } else {
                    console.error(`Failed to send webhook. Status: ${response.status}`);
                    return { success: false, error: `HTTP ${response.status}` };
                }
            } catch (error: any) {
                if (attempt < retryCount - 1) {
                    console.warn(`Failed to send webhook. Retrying in 5 seconds. Error: ${error.message}`);
                    await new Promise(resolve => setTimeout(resolve, 5000));
                } else {
                    console.error(`Failed to send webhook after ${retryCount} attempts. Error: ${error.message}`);
                    return { success: false, error: error.message };
                }
            }
        }

        return { success: false, error: 'Max retries exceeded' };
    }

    /**
     * Query webhook (expects response)
     */
    async queryWebhook(
        data: any, 
        retryCount: number = 3, 
        timeout: number = 10000
    ): Promise<any> {
        const url = process.env.FASTAPI_API_URL || 'http://localhost:8000/api/ai/receive-webhook';
        
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'User-Agent': 'Fleet Dispatch Assistant',
            'x-webhook-secret': this.secretKey
        };
        
        const payload: WebhookPayload = {
            event: 'query',
            data
        };

        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), timeout);

            const response = await fetch(url, {
                method: 'POST',
                headers,
                body: JSON.stringify(payload),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            if (response.ok) {
                return await response.json();
            } else {
                throw new Error(`Failed to query webhook. Status: ${response.status}; Response: ${await response.text()}`);
            }
        } catch (error: any) {
            throw new Error(`Failed to query webhook. Error: ${error.message}`);
        }
    }
}

export { WebhookSender };